modules/ln_core/utf8string.scm

```diff
@@ -177,53 +177,86 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+      #xfffd)
+    (define (select key)
+      (set!
+       handler
+       (if (procedure? key)
+           key
+           (case key
+             ((replace #f) utf8string->unicode:on-encoding-error-replace)
+             ((log+replace) utf8string->unicode:on-encoding-error-log+replace)
+             (else utf8string->unicode:on-encoding-error-raise-error)))))
+    (select 'error)
+    (lambda args
+      (if (null? args) handler (select (car args))))))
+
+(define (utf8string->unicode str #!optional (encoding-error #t))
+  (define (on-encoding-error i)
+    (cond
+     ((not encoding-error) #xfffd) ;; deliver replacement charater
+     ((procedure? encoding-error) (encoding-error str i)) ;; delegate to caller
+     (else ((utf8string->unicode:on-encoding-error) str i)))) ;; delegate to registered
+  (let next-char ((i 0))
+    (if (fx>= i (string-length str)) '()
+        (let* ((c (string-ref str i))
+               (c1 (char->integer c)))
+                (receive (size m1)
+                    (cond
+                     ((fx< c1 #x80) (values 1 #x7f))
+                     ((fx< c1 #xe0) (values 2 #x0f))
+                     ((fx< c1 #xf0) (values 3 #x0f))
+                     ((fx< c1 #xf8) (values 4 #x07))
+                     ((fx< c1 #xfc) (values 5 #x03))
+                     (else (values 6 #x01)))
+                  (if (fx= size 1)
+                      (cons c1 (next-char (fx+ i 1)))
+                      (let ((limit (fx+ i size)))
+                        (let subc ((j (fx+ i 1)) (r (bitwise-and c1 m1)))
+                          (cond
+                           ((fx>= j limit) (cons r (next-char limit)))
+                           ((fx>= j (string-length str)) (list (on-encoding-error j)))
+                           (else
+                            (let ((cc (char->integer (string-ref str j))))
+                              (if (or (fx< cc #x80) (fx>= cc #xc0))
+                                  (let ((r (on-encoding-error j))) ;; 1st force handling
+                                    (cons r (next-char j)))
+                                  (subc
+                                   (fx+ j 1)
+                                   (bitwise-ior
+                                    (arithmetic-shift r 6)
+                                    (bitwise-and cc #x3F)))))))))))))))
+
+(define (utf8string->unicode/fallback str #!key (fallback #f))
+  ;; A version, which replaces string causing convertion erros with a
+  ;; hopefuly safe fallback.
+  (let ((use-fallback (list 0))) ;; just a well known value
+    (with-exception-catcher
+     (lambda (exn)
+       (if (eq? exn use-fallback)
+           (case fallback
+             ((#f) (map char->integer (string->list str)))
+             (else (fallback str)))
+           (handle-replloop-exception exn)))
+     (lambda () (utf8string->unicode str (lambda (str i) (raise use-fallback)))))))
+
+(define unicode->utf8string unicode->utf8string)
```

Comment:
You define it to itself and declare it below - having just resolved order issues with the removal of (block) this doesn't quite make sense to me.